src/lib/y2storage/clients/partitions_proposal.rb

```diff
@@ -0,0 +1,145 @@
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "yast"
+require "y2storage"
+require "y2storage/actions_presenter"
+require "y2storage/dialogs/proposal"
+require "installation/proposal_client"
+
+Yast.import "Popup"
+
+module Y2Storage
+  module Clients
+    # Proposal client to show the list of storage actions.
+    #
+    # This client is used in autoyast summary dialog, where a new
+    # client instance is created with each event that requires to
+    # update the summary dialog content.
+    #
+    # To manage collapsing/expanding subvolumes list it is necessary
+    # to save the previous state of that list, but it is not possible
+    # to do that in the client instance itself because a new client
+    # instance is created each time.
+    #
+    # To solve that, instance class attributes are used. PartitionsProposal
+    # class has an internal state that is updated with each object creation.
+    #
+    # @see PartitionsProposal.update_state
+    class PartitionsProposal < ::Installation::ProposalClient
+      include Yast::Logger
+
+      def initialize
+        textdomain "storage-ng"
+        ensure_proposed unless storage_manager.staging_changed?
+        self.class.update_state
+      end
+
+      def make_proposal(_attrs)
+        {
+          "preformatted_proposal" => actions_presenter.to_html,
+          "links"                 => actions_presenter.events,
+          "language_changed"      => false
+        }
+      end
+
+      def ask_user(param)
+        event = param["chosen_id"]
+
+        if actions_presenter.can_handle?(event)
+          actions_presenter.update_status(event)
+          result = :again
+        else
+          Yast::Report.Warning(_("This is not enabled at this moment (event: %s)") % event)
+          log.warn("WARNING: impossible to manage event #{event}")
+          result = :back
+        end
+
+        { "workflow_sequence" => result }
+      end
+
+      def description
+        {
+          "id"              => "partitions",
+          "rich_text_title" => _("Partitioning"),
+          "menu_title"      => _("&Partitioning")
+        }
+      end
+
+    private
+
+      class << self
+        attr_accessor :staging_revision
+        attr_accessor :actions_presenter
+
+        # Updates internal class state when it is necessary.
+        #
+        # A new actions presenter is created when the current staging revision
+        # is different to the last saved revision.
+        #
+        # @see ActionsPresenter
+        def update_state
+          storage_manager = StorageManager.instance
+          return if staging_revision == storage_manager.staging_revision
+
+          self.staging_revision = storage_manager.staging_revision
+
+          staging = storage_manager.y2storage_staging
+          actiongraph = staging ? staging.actiongraph : nil
```

Comment:
First of all, I miss a RSpec example for `ActionsPresenter` explaining what is returned by `#to_html` when the object is initialized with `nil`. I assume `#to_html` returns the empty string or an empty `<p>`.

That being said, is that the expected output of `make_proposal` in that case or show this client show some error message (even an error blocking the installation to continue)?